### PR TITLE
exporters

### PR DESCRIPTION
**Pull request by micwill755**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant and/or add your own.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

# Checklist:

- [ ] My code follows the style guidelines of this project (You can use the linters)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added the relevant labels to my PR in so that relevant reviewers are notified

---

<sub>Stack created with <a href="https://github.com/github/gh-stack">GitHub Stacks CLI</a> • <a href="https://gh.io/stacks-feedback">Give Feedback 💬</a></sub>

### CODE REVIEW

File: examples/dynamo/run_groot_export.py
Comment (narendasan) [inline examples/dynamo/run_groot_export.py:1]: You can just put the example of the exporter tool with the tool unless you want them rendered. Then we need to use the sphinx gallery format

Comment (narendasan) [inline examples/dynamo/run_groot_export.py:1]: Do you have the inference side of these examples? (Ideally both python and C++ deployments)

File: py/torch_tensorrt/hf/exporters/models/common/__init__.py
Comment (narendasan) [inline py/torch_tensorrt/hf/exporters/models/common/__init__.py:1]: Can we start this out of tree for now like //tools/hf_exporters?

File: py/torch_tensorrt/hf/exporters/models/common/helpers.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    """inputs_embeds, rope, ctx, kv_start, last_token_ids, ds_stack, *kvs."""
+    decoder = getattr(language, "model", language)
+    cfg = language.config
+    bsz, prompt_len, hidden = inputs_embeds.shape
+    seq_len = int(seq_len or prompt_len)
+    num_layers = len(decoder.layers)
+    num_kv = int(cfg.num_key_value_heads)
+    head_dim = int(getattr(cfg, "head_dim", cfg.hidden_size // cfg.num_attention_heads))
+    try:
+        from torch_tensorrt.hf.exporters.rope import make_rope_rotary_cos_sin
```
Comment (narendasan) [inline py/torch_tensorrt/hf/exporters/models/common/helpers.py:28]: Why do we need to guard around code we ship?

File: py/torch_tensorrt/hf/exporters/plugin/plugin_converter.py [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+        ctx.net.add_plugin_v3(inputs, [], plugin)
+        if _plugin_is_v3(plugin)
+        else ctx.net.add_plugin_v2(inputs, plugin)
+    )
+    layer.name = name
+    return layer
+
+
+@dynamo_tensorrt_converter(
+    torch.ops.trt.attention_plugin.default,
```
Comment (narendasan) [inline py/torch_tensorrt/hf/exporters/plugin/plugin_converter.py:52]: make this namespace torch.ops.tensorrt_edge_llm.*

File: py/torch_tensorrt/hf/exporters/plugin/plugin_converter.py [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+        shuffle_layer = ctx.net.add_shuffle(inputs[kv_cache_start_idx_input_idx])
+        shuffle_layer.reshape_dims = (inputs[kv_cache_start_idx_input_idx].shape[0],)
+        inputs[kv_cache_start_idx_input_idx] = shuffle_layer.get_output(0)
+
+    layer = _add_plugin_layer(ctx, inputs, plugin, name)
+    return layer.get_output(0), layer.get_output(1)
+
+
+@dynamo_tensorrt_converter(
+    torch.ops.trt.vit_attention_plugin.default,
```
Comment (narendasan) [inline py/torch_tensorrt/hf/exporters/plugin/plugin_converter.py:133]: Same here

File: py/torch_tensorrt/hf/exporters/plugin/plugin_converter.py [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+            else tensor
+        )
+        if not getattr(trt_tensor, "name", None):
+            trt_tensor.name = tensor_name
+        inputs.append(trt_tensor)
+    return inputs
+
+
+@dynamo_tensorrt_converter(
+    torch.ops.trt.causal_conv1d.default,
```
Comment (narendasan) [inline py/torch_tensorrt/hf/exporters/plugin/plugin_converter.py:211]: Here as well

File: py/torch_tensorrt/hf/exporters/plugin/plugin_converter.py [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+
+    inputs = _as_plugin_inputs(
+        ctx, [hidden_states, weight, bias, conv_state, context_lengths], name
+    )
+    layer = _add_plugin_layer(ctx, inputs, plugin, name)
+    return layer.get_output(0), layer.get_output(1)
+
+
+@dynamo_tensorrt_converter(
+    torch.ops.trt.update_ssm_state.default,
```
Comment (narendasan) [inline py/torch_tensorrt/hf/exporters/plugin/plugin_converter.py:248]: ""

File: py/torch_tensorrt/hf/exporters/plugin/plugin_converter.py [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+    if plugin is None:
+        raise RuntimeError(f"Failed to create {plugin_name}")
+
+    inputs = _as_plugin_inputs(ctx, tensors, name)
+    layer = _add_plugin_layer(ctx, inputs, plugin, name)
+    return layer.get_output(0)
+
+
+@dynamo_tensorrt_converter(
+    torch.ops.trt.nvfp4_moe_plugin.default,
```
Comment (narendasan) [inline py/torch_tensorrt/hf/exporters/plugin/plugin_converter.py:348]: ""

File: py/torch_tensorrt/hf/exporters/utils.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+from __future__ import annotations
+
+import os
+from pathlib import Path
+from typing import Any
+
+import torch
+
+_THOR_CUDA_LIB = Path("/usr/local/cuda-13.0/thor/targets/aarch64-linux/lib")
```
Comment (narendasan) [inline py/torch_tensorrt/hf/exporters/utils.py:11]: We should have some more robust discovery method

**Review by narendasan — COMMENTED** [review]
This is cool, for sure much more clean. Do you have like a guide on how to bring up your own model?

File: examples/dynamo/run_groot_export.py
Comment (micwill755) [reply examples/dynamo/run_groot_export.py:1]: Sure, I added another PR on this stack with rough docs, still need to add inference instructions

File: py/torch_tensorrt/hf/exporters/utils.py (hunk 7 above)
Comment (micwill755) [reply py/torch_tensorrt/hf/exporters/utils.py:11]: I removed this, it was stale code left in the repo
